### PR TITLE
Bleak Fork Upstream Update

### PR DESCRIPTION
**Pull request by pmessan**
This PR proposes an update to use the proglove_next branch as our updated bleak fork. A few patches were added including resume and suspend methods to the BleakScannerBlueZDBus class, and setup for Cloudsmith.

### CODE REVIEW

File: bleak/backends/bluezdbus/scanner.py [hunk 1]
```diff
@@ -141,53 +141,69 @@ async def start(self):
 
         logger.debug(f"cached devices: {self._cached_devices}")
 
-        # Apply the filters
+        await self.resume()
+
+    async def stop(self):
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: why did you move it here? It makes the diff impossible to understand.

File: bleak/backends/bluezdbus/scanner.py [hunk 2]
```diff
@@ -141,53 +141,69 @@ async def start(self):
...
+
+        # Try to disconnect the System Bus.
+        try:
+            self._bus.disconnect()
+        except Exception as e:
+            logger.error("Attempt to disconnect system bus failed: {0}".format(e))
+
+        self._bus = None
+
+    # add suspend()
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: what is that comment?

File: bleak/backends/bluezdbus/scanner.py [hunk 3]
```diff
@@ -141,53 +141,69 @@ async def start(self):
...
+                member="StopDiscovery",
             )
         )
         assert_reply(reply)
 
-        # Start scanning
+    # add resume()
+    async def resume(self):
+        # if self.is_scanning:
+        #     return
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: what is that dead code?

File: bleak/backends/bluezdbus/scanner.py [hunk 4]
```diff
@@ -141,53 +141,69 @@ async def start(self):
...
-                member="StartDiscovery",
-            )
+        Message(
+            destination=defs.BLUEZ_SERVICE,
+            path=self._adapter_path,
+            interface=defs.ADAPTER_INTERFACE,
+            member="SetDiscoveryFilter",
+            signature="a{sv}",
+            body=[self._filters],
+                )
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: why did the indentation change?

File: bleak/backends/bluezdbus/scanner.py [hunk 5]
```diff
@@ -141,53 +141,69 @@ async def start(self):
...
                 destination=defs.BLUEZ_SERVICE,
                 path=self._adapter_path,
                 interface=defs.ADAPTER_INTERFACE,
-                member="StopDiscovery",
+                member="StartDiscovery",
             )
         )
         assert_reply(reply)
+        # except RemoteError as e:
+        #     logger.error("Stop discovery failed: {0}".format(e))
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: what is that dead code?

File: bleak/backends/bluezdbus/scanner.py [hunk 6]
```diff
@@ -141,53 +141,69 @@ async def start(self):
...
-            await remove_match(self._bus, rule)
-        self._rules.clear()
-        self._bus.remove_message_handler(self._parse_msg)
-
-        # Try to disconnect the System Bus.
-        try:
-            self._bus.disconnect()
-        except Exception as e:
-            logger.error("Attempt to disconnect system bus failed: {0}".format(e))
+        # self.is_scanning = True
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: why was this added?

**Review by arthur-proglove — CHANGES_REQUESTED** [review]
Are you sure you added all the important commits? Only 2 looks very small to me.

- From what I see there are at least 3 commits on `setup.py` that should make it here.
- Do we have a new method to get the MTU in the upstream bleak? (1f06df9171eb99e55a1a9c0b05d74c9533a9f92e)
- Do we still support the skip discovery flag? (2115ef0079bd07018a45a144971e65667e53eae2)
- What about the cache? (8e9678a10d0245e6dbe11df07ad56ab4ad21d0cd, 93e783f1b16f91fb822724138c90203dd10bf74b, 43ecf105d838342e93215cb065bc9733dcbe476f)
- Do we still check for BlueZ crashses (cad63b0da22bf65b2a81435bc33196fe92e2d8c1) ?

File: bleak/backends/bluezdbus/scanner.py (hunk 2 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Removed

File: bleak/backends/bluezdbus/scanner.py (hunk 3 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Removed

File: bleak/backends/bluezdbus/scanner.py (hunk 4 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Fixed

File: bleak/backends/bluezdbus/scanner.py (hunk 5 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Removed

File: bleak/backends/bluezdbus/scanner.py (hunk 6 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Removed

File: bleak/backends/bluezdbus/scanner.py (hunk 1 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: Reorganised it now

File: bleak/backends/bluezdbus/scanner.py [hunk 7]
```diff
@@ -189,6 +196,43 @@ async def stop(self):
...
+                path=self.adapter_path,
+                member="Set",
+                interface=defs.PROPERTIES_INTERFACE,
+                destination=defs.BLUEZ_SERVICE,
+                signature="ssv",
+                body=[defs.ADAPTER_INTERFACE, 'Powered', Variant("b", True)],
+            )
+        )
+    async def _restart(self):
+        """Stop and restart the discovery."""
```
Comment (pmessan) [inline bleak/backends/bluezdbus/scanner.py]: I'm a little lost as to where this is called exactly.

**Comment by pmessan** [comment]
> Are you sure you added all the important commits? Only 2 looks very small to me.
> 
> * From what I see there are at least 3 commits on `setup.py` that should make it here.

Looked at it, and added a commit with the releva.nt changes

> * Do we have a new method to get the MTU in the upstream bleak? ([1f06df9](https://github.com/workaroundgmbh/bleak/commit/1f06df9171eb99e55a1a9c0b05d74c9533a9f92e))

In the upstream branch, There's already a [method ](https://github.com/workaroundgmbh/bleak/blob/d6c208f15bf3ad84338d6904dc0a2ee1242e27a9/bleak/backends/bluezdbus/client.py#L519)called `_acquire_mtu` which does this.
 
> * Do we still support the skip discovery flag? ([2115ef0](https://github.com/workaroundgmbh/bleak/commit/2115ef0079bd07018a45a144971e65667e53eae2))

I think we don't need to any more because the upstream `get_services` [method](https://github.com/workaroundgmbh/bleak/blob/d6c208f15bf3ad84338d6904dc0a2ee1242e27a9/bleak/backends/bluezdbus/client.py#L585) only returns the services if they are resolved, hence fixing the problem of bluez crashing if we don't wait for the services to be resolved.

> * What about the cache? ([8e9678a](https://github.com/workaroundgmbh/bleak/commit/8e9678a10d0245e6dbe11df07ad56ab4ad21d0cd), [93e783f](https://github.com/workaroundgmbh/bleak/commit/93e783f1b16f91fb822724138c90203dd10bf74b), [43ecf10](https://github.com/workaroundgmbh/bleak/commit/43ecf105d838342e93215cb065bc9733dcbe476f))

I added the patches for the cache, since I did not find any pre-exisiting code that checked if the cache was enabled. I think those patches are still relevant.

> * Do we still check for BlueZ crashses ([cad63b0](https://github.com/workaroundgmbh/bleak/commit/cad63b0da22bf65b2a81435bc33196fe92e2d8c1)) ?

Agreed to not focus on this for now

**Comment by pmessan** [comment]
> > Are you sure you added all the important commits? Only 2 looks very small to me.
> > 
> > * From what I see there are at least 3 commits on `setup.py` that should make it here.
> > * Do we have a new method to get the MTU in the upstream bleak? ([1f06df9](https://github.com/workaroundgmbh/bleak/commit/1f06df9171eb99e55a1a9c0b05d74c9533a9f92e))
> 
> In the upstream branch, There's already a method called `_acquire_mtu` which does this.
> 
> > * Do we still support the skip discovery flag? ([2115ef0](https://github.com/workaroundgmbh/bleak/commit/2115ef0079bd07018a45a144971e65667e53eae2))
> 
> I think we don't need to any more because the upstream `get_services` [method](https://github.com/workaroundgmbh/bleak/blob/d6c208f15bf3ad84338d6904dc0a2ee1242e27a9/bleak/backends/bluezdbus/client.py#L585) only returns the services if they are resolved, hence fixing the problem of bluez crashing if we don't wait for the services to be resolved.
> 
> > * What about the cache? ([8e9678a](https://github.com/workaroundgmbh/bleak/commit/8e9678a10d0245e6dbe11df07ad56ab4ad21d0cd), [93e783f](https://github.com/workaroundgmbh/bleak/commit/93e783f1b16f91fb822724138c90203dd10bf74b), [43ecf10](https://github.com/workaroundgmbh/bleak/commit/43ecf105d838342e93215cb065bc9733dcbe476f))
> 
> I added the patches for the cache, since I did not find any pre-exisiting code that checked if the cache was enabled. I think those patches are still relevant.
> 
> > * Do we still check for BlueZ crashses ([cad63b0](https://github.com/workaroundgmbh/bleak/commit/cad63b0da22bf65b2a81435bc33196fe92e2d8c1)) ?

Agreed to not focus on this for now

File: bleak/backends/bluezdbus/client.py [hunk 8]
```diff
@@ -84,6 +87,28 @@ def __init__(self, address_or_ble_device: Union[BLEDevice, str], **kwargs):
         self._hides_battery_characteristic = check_bluez_version(5, 48)
         self._hides_device_name_characteristic = check_bluez_version(5, 48)
 
+
+    def _is_cache_enabled() -> bool:
+        global _cache_enabled
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/client.py]: this could be a class member

File: bleak/backends/bluezdbus/scanner.py [hunk 9]
```diff
@@ -143,6 +144,42 @@ async def start(self):
 
         await self.resume()
 
+    async def resume(self):
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py:146]: This does not really match the commit description?

**Review by arthur-proglove — CHANGES_REQUESTED** [review]
You did not cherry pick the commits but created your own for some changes that could just have been taken from the git history (setup.py and .travis.yml). I updated this PR with the proper commits.

File: bleak/backends/bluezdbus/scanner.py [hunk 10]
```diff
@@ -189,6 +196,43 @@ async def stop(self):
...
+            Message(
+                path=self.adapter_path,
+                member="Set",
+                interface=defs.PROPERTIES_INTERFACE,
+                destination=defs.BLUEZ_SERVICE,
+                signature="ssv",
+                body=[defs.ADAPTER_INTERFACE, 'Powered', Variant("b", True)],
+            )
+        )
+    async def _restart(self):
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py]: this is not even called

File: bleak/backends/bluezdbus/scanner.py [hunk 11]
```diff
@@ -165,7 +169,7 @@ async def start(self):
         )
         assert_reply(reply)
 
-    async def stop(self):
+    async def suspend(self):
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/scanner.py:171]: so `stop` just became `suspend`? :-D

**Review by arthur-proglove — CHANGES_REQUESTED** [review]

File: bleak/backends/bluezdbus/scanner.py (hunk 11 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py:171]: No, `suspend` is called inside of stop, like how it was done [here](https://github.com/workaroundgmbh/bleak/blob/859f7b966bbabdff29908e07e534f09e8b49ee40/bleak/backends/bluezdbus/discovery.py#L256)

File: bleak/backends/bluezdbus/scanner.py (hunk 9 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py:146]: I think it does though, it was added in [this commit](https://github.com/workaroundgmbh/bleak/pull/33/commits/61c12e54b5750bae751d78889c1bfb0d5d953497#diff-cc469ec95c55bf20bb32d39bf09be9cec2dfc6820e40e857367f8b49915cddcdR175)

File: bleak/backends/bluezdbus/client.py (hunk 8 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/client.py]: Fixed

File: bleak/backends/bluezdbus/scanner.py (hunk 10 above)
Comment (pmessan) [reply bleak/backends/bluezdbus/scanner.py]: removed

**Comment by arthur-proglove** [comment]
Okay looks good to me, we're just missing `_acquire_mtu` to be exposed. Somehow there is a property `mtu_size` but since `_acquire_mtu` is not called anywhere `_mtu_size` is not set. 
I think `_acquire_mtu` should be called `once` after being connected, maybe add one more commit with this patch?

**Comment by pmessan** [comment]
> Okay looks good to me, we're just missing `_acquire_mtu` to be exposed. Somehow there is a property `mtu_size` but since `_acquire_mtu` is not called anywhere `_mtu_size` is not set. I think `_acquire_mtu` should be called `once` after being connected, maybe add one more commit with this patch?

Patch added!

File: bleak/backends/bluezdbus/client.py [hunk 12]
```diff
@@ -340,6 +340,8 @@ async def connect(self, **kwargs) -> bool:
             # Get all services. This means making the actual connection.
             await self.get_services()
 
+            await self._acquire_mtu()
```
Comment (arthur-proglove) [inline bleak/backends/bluezdbus/client.py:343]: you're not exposing it ;-) you're calling it

Comment (pmessan) [reply bleak/backends/bluezdbus/client.py:343]: Reworded the commit 😆

**Review by arthur-proglove — APPROVED** [review]
LGTM :-)

**Comment by pmessan** [comment]
Apparently I'm not authorized to merge 😆 
How do I fix that?

**Comment by arthur-proglove** [comment]
> Apparently I'm not authorized to merge laughing How do I fix that?

let me check

**Review by arthur-proglove — APPROVED** [review]
